### PR TITLE
Fixed ONNXImporter

### LINKED ISSUE #16370: DNN: Required argument "split" not found into dictionary
##### System information (version)
- OpenCV => 4.2.0
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2019

##### Detailed description
I'm trying to inference classification model trained on [Custom Vision](https://www.customvision.ai/) of Azure Cognitive Services using OpenCV DNN module.  
Custom Vision can download trained ONNX model. The following error occurs when read net from it.  

```
OpenCV: terminate handler is called! The last OpenCV error is:
OpenCV(4.2.0) Error: Requested object was not found (Required argument "split" not found into dictionary) in cv::dnn::dnn4_v20191202::Dict::get, file C:\opencv-4.2.0\modules\dnn\include\opencv2/dnn/dnn.inl.hpp, line 349
```

This is part of model visualize image using Netron.
2020-01-17_10h51_43
(Other model visualize image is [here](https://user-images.githubusercontent.com/816705/72579941-e8282000-391d-11ea-8bee-fbfc2989c6e3.png). All model visualize image is [here](https://user-images.githubusercontent.com/816705/72579836-98e1ef80-391d-11ea-9b8e-60a2aed95f35.png).)

What do you think? Thanks,

##### Steps to reproduce
This issue can be reproduced with the following code: 

```cpp
const std::string model = "./model.onnx";
cv::dnn::Net net = cv::dnn::readNet( model ); // error
```

### PR DESCRIPTION
**Pull request by czgdp1807**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/701

resolves #16370 

### This pullrequest changes

This Pull Request aims to fix ONNXImporter. Currently, there are few bugs in processing of various `layer_type` from `.onnx` files. The referenced issue reports a bug in `Split` layer. Though I have observed that there are more bugs. I have listed them below as tasks. I will update them every commit.

- [x] "Split" layer fixed.
- [x] Tests added

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.1.0:16.04
build_image:Custom Win=openvino-2020.1.0
build_image:Custom Mac=openvino-2020.1.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by dkurt** [comment]
Thank you!

Can you please for each fixed case provide a simple one-layer test by the following script: https://github.com/opencv/opencv_extra/blob/master/testdata/dnn/onnx/generate_onnx_models.py?

**Comment by czgdp1807** [comment]
Hi, @dkurt
I was trying to import the ONNX model after making a fix to `split` layer. However, other layers were not processed correctly like, when `layer_type` is `Constant`, `attribute_proto.t()` is returning an empty tensor. I suspect there is a problem with `.onnx` file shared in https://github.com/opencv/opencv/issues/16370#issuecomment-575441365 as the visualization of the model doesn't have any `Constant` layer but the `graph_proto` is showing to have it and that too with `string` data type which is not supported by `getMatFromTensors`. 
I am still trying to figure out what's not working correctly. I will add the tests once the model is processed correctly. Thank you.

File: modules/dnn/test/test_onnx_importer.cpp [hunk 1]
```diff
@@ -386,6 +386,11 @@ TEST_P(Test_ONNX_layers, ReduceL2)
     testONNXModels("reduceL2");
 }
 
+TEST_P(Test_ONNX_layers, Split)
+{
+    testONNXModels("split");
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp]: There is no `split` tests in this PR. There are `split_1`, `split_2` and `split_3`. See https://github.com/opencv/opencv_extra/pull/701

```
[ RUN      ] Test_ONNX_layers.Split/0, where GetParam() = OCV/CPU
unknown file: Failure
C++ exception with description "OpenCV(3.4.9-dev) /build/precommit_linux64/3.4/opencv/modules/ts/src/ts.cpp:1052: error: (-2:Unspecified error) OpenCV tests: Can't find required data file: dnn/onnx/models/split.onnx in function 'findData'
" thrown in the test body.
[  FAILED  ] Test_ONNX_layers.Split/0, where GetParam() = OCV/CPU (0 ms)
https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/24578/steps/test_dnn/logs/stdio
```

**Comment by czgdp1807** [comment]
Update:
Hi,
I will be having my mid semester examinations till 7th February(IST). So the work on this may get slowed down. Please do not close this PR until then. I am on it and I will put efforts to complete it from 7th Feb. Thank you and apologies.

**Comment by czgdp1807** [comment]
Hi, 
I am unable to figure out how to run the tests for this PR locally.
I tried running, ` python3 generate_onnx_models.py` under `/opencv_project/opencv_extra/testdata/dnn/onnx` but got, 
```
maxpooling input has sizes torch.Size([1, 3, 10, 9])
maxpooling output has sizes torch.Size([1, 3, 3, 3])

Segmentation fault (core dumped)
```
Do we have to build the code after generating models or they are generated automatically during the build?
I am unable to see logs at https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/24598/steps/test_dnn

**Comment by alalek** [comment]
> unable to see logs

These logs were expired.
I rescheduled new builds, please check.

**Comment by czgdp1807** [comment]
Thanks @alalek 
After fixing a probable bug in `Constant` layer, I observed that `ArgMax` hasn't been handled. I got the following exception,
```cpp
 what():  OpenCV(3.4.9-dev) /home/czgdp1807/opencv_project/opencv/modules/dnn/src/dnn.cpp:561: error: (-2:Unspecified error) Can't create layer "LabelId" of type "ArgMax" in function 'getLayerInstance'
```
Should I add a new file under `layers` for this and some more uncovered layers(`ArgMin`)?

**Comment by dkurt** [comment]
@czgdp1807, Can you please dump a subgraph from the model which includes ArgMin node? Just to understand which layer it tries to perform.

**Comment by czgdp1807** [comment]
I have attached a screenshot below of the last few layers. The `ArgMax` isn't created as the case for it isn't covered as far as I am able to figure out.
Screenshot from 2020-02-09 00-41-46

**Comment by dkurt** [comment]
Got it, thanks! This is training only nodes so author should exclude them from deployment model (you may try to load and save model but before export call `.eval()` method or specify SoftMax as output.

So let's focus on Split layer but ask author to try to remove training nodes.

**Comment by czgdp1807** [comment]
Thanks. IMO, this could be one of the ways as the last few layers aren't making any difference to the result, though the interpretation would be different. 
I will drop the last commit from here and will add test files for `Split` layer in my other PR.

**Comment by czgdp1807** [comment]
Please restart the tests here as I have made an update to `opencv_extra`.
Thanks.

**Comment by czgdp1807** [comment]
> Please restart the tests here as I have made an update to `opencv_extra`.
> Thanks.

+1

**Comment by alalek** [comment]
Amend commit (no changes are required, just update commit date) and push "new" changes to this PR:
```
git commit --amend
git push <remote> <branch> --force
```

**Comment by czgdp1807** [comment]
Done. Locally the tests were passing. Let's see how it goes here.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -398,7 +398,6 @@ void ONNXImporter::populateNet(Net dstNet)
         std::string layer_type = node_proto.op_type();
         layerParams.type = layer_type;
 
-
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp]: Please avoid unrelated changes (extra lines addings or removals)

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 3]
```diff
@@ -483,17 +482,25 @@ void ONNXImporter::populateNet(Net dstNet)
          }
         else if (layer_type == "Split")
         {
-            DictValue splits = layerParams.get("split");
-            const int numSplits = splits.size();
-            CV_Assert(numSplits > 1);
+            if(layerParams.has("split"))
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp]: Please keep a single whitespace  after if:
```cpp
if (
```

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 4]
```diff
@@ -483,17 +482,25 @@ void ONNXImporter::populateNet(Net dstNet)
...
+                    slicePoints[i] = slicePoints[i - 1] + splits.get<int>(i - 1);
+                }
+                layerParams.set("slice_point", DictValue::arrayInt(&slicePoints[0], slicePoints.size()));
+                layerParams.type = "Slice";
+            }
+            else
             {
-                slicePoints[i] = slicePoints[i - 1] + splits.get<int>(i - 1);
+                layerParams.set("num_split", node_proto.output_size());
+                layerParams.type = "Slice";
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp]: `layerParams.type = "Slice";` is the same for both cases. So let's just move it outside if-else block

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (czgdp1807) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Ah! Sorry. I will remove it.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 4 above)
Comment (czgdp1807) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Sure. I will make this change.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Never mind. Just to make this PR perfect :)

**Comment by czgdp1807** [comment]
I have applied the comments and reviews. Let me know if anything else is to be done. Tests should pass here.

**Comment by czgdp1807** [comment]
The tests passed. Is it good to go?

**Comment by dkurt** [comment]
@czgdp1807, some of test from Inference Engine backend are failed:
```
[ RUN      ] Test_ONNX_layers.Split/1, where GetParam() = NGRAPH/CPU
/build/precommit_custom_mac/3.4/opencv/modules/dnn/test/test_common.impl.hpp:66: Failure
Expected: (normL1) <= (l1), actual: 0.5 vs 1e-05
/build/precommit_custom_mac/3.4/opencv/modules/dnn/test/test_common.impl.hpp:69: Failure
Expected: (normInf) <= (lInf), actual: 1 vs 0.0001
FALLBACK: Layer [Slice]:[1] is expected to has backend implementation
/build/precommit_custom_mac/3.4/opencv/modules/dnn/test/test_common.impl.hpp:66: Failure
Expected: (normL1) <= (l1), actual: 0.5 vs 1e-05
/build/precommit_custom_mac/3.4/opencv/modules/dnn/test/test_common.impl.hpp:69: Failure
Expected: (normInf) <= (lInf), actual: 1 vs 0.0001
FALLBACK: Layer [Slice]:[1] is expected to has backend implementation
/build/precommit_custom_mac/3.4/opencv/modules/dnn/test/test_common.impl.hpp:66: Failure
Expected: (normL1) <= (l1), actual: 0.5 vs 1e-05
/build/precommit_custom_mac/3.4/opencv/modules/dnn/test/test_common.impl.hpp:69: Failure
Expected: (normInf) <= (lInf), actual: 1 vs 0.0001
FALLBACK: Layer [Slice]:[1] is expected to has backend implementation
/build/precommit_custom_mac/3.4/opencv/modules/dnn/test/test_common.impl.hpp:66: Failure
Expected: (normL1) <= (l1), actual: 0.5 vs 1e-05
/build/precommit_custom_mac/3.4/opencv/modules/dnn/test/test_common.impl.hpp:69: Failure
Expected: (normInf) <= (lInf), actual: 1 vs 0.0001
FALLBACK: Layer [Slice]:[1] is expected to has backend implementation
```
I can take a look and tell how to fix it

**Comment by czgdp1807** [comment]
How to run the failing tests locally, as no test was failing on, `./opencv_test_dnn`.

**Comment by dkurt** [comment]
@czgdp1807, it's optional backend for deep learning module: https://github.com/opencv/opencv/wiki/Intel%27s-Deep-Learning-Inference-Engine-backend (OpenVINO). Anyway there are other tests for split which are failed as well. So let's skip it for now but fix them all separately.

**Comment by czgdp1807** [comment]
Sure. Can we open a separate issue for the remaining failures in `Split` layer? I will work on it in a different PR.

**Comment by dkurt** [comment]
Yep, let's open a separate issue to not to miss. Thanks!

**Comment by czgdp1807** [comment]
I have updated the code and opened a separate issue. Please let me know, if any more update is to be done in this PR.

**Pull request merged**